### PR TITLE
Cli updates to release

### PR DESCRIPTION
**Pull request by sapinb**
## Description

Cherrypick cli related changes from main to release

### Type of Change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature/Enhancement (non-breaking change which adds functionality or enhances an existing one)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update
-   [ ] Refactor
-   [ ] New or updated tests

## Checklist

-   [x] I have performed a self-review of my code.
-   [ ] I have commented my code where necessary.
-   [ ] I have updated the documentation if needed.
-   [x] My changes do not introduce new warnings.
-   [ ] I have added tests that prove my changes are effective or that my feature works.
-   [x] New and existing tests pass with my changes.

### CODE REVIEW

**Comment by storopoli** [comment]
We also need to update the private-docs `cargo install` command

**Review by storopoli — CHANGES_REQUESTED** [review]
We can bring #423 it's squash merged

**Comment by storopoli** [comment]
I've brought #423 in a743f65254930556e5662d0e339e1c44347c732a.
Let's wait for CI

**Review by storopoli — DISMISSED** [review]
ACK a743f65254930556e5662d0e339e1c44347c732a

**Comment by storopoli** [comment]
Done in #437

File: bin/strata-cli/src/cmd/balance.rs [hunk 1]
```diff
@@ -23,13 +24,14 @@ pub struct BalanceArgs {
...
+pub async fn balance(args: BalanceArgs, seed: Seed, settings: Settings) {
     let term = Term::stdout();
     let network_type = net_type_or_exit(&args.network_type, &term);
 
     if let NetworkType::Signet = network_type {
-        let mut l1w = SignetWallet::new(&seed, NETWORK).unwrap();
-        l1w.sync(&esplora).await.unwrap();
+        let mut l1w =
+            SignetWallet::new(&seed, settings.network, settings.signet_backend.clone()).unwrap();
+        l1w.sync().await.unwrap();
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/balance.rs:34]: Fix this `.unwrap()`.

File: bin/strata-cli/src/cmd/balance.rs [hunk 2]
```diff
@@ -42,11 +44,14 @@ pub async fn balance(args: BalanceArgs, seed: Seed, settings: Settings, esplora:
     }
 
     if let NetworkType::Strata = network_type {
-        let l2w = StrataWallet::new(&seed, &settings.l2_http_endpoint).unwrap();
+        let l2w = StrataWallet::new(&seed, &settings.strata_endpoint).unwrap();
         let _ = term.write_line("Getting balance...");
         let balance = l2w.get_balance(l2w.default_signer_address()).await.unwrap();
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/balance.rs:49]: Fix these `.unwrap()`s.

File: bin/strata-cli/src/cmd/deposit.rs [hunk 3]
```diff
@@ -16,34 +16,45 @@ use strata_bridge_tx_builder::constants::MAGIC_BYTES;
...
-    let mut l1w = SignetWallet::new(&seed, NETWORK).unwrap();
-    let l2w = StrataWallet::new(&seed, &settings.l2_http_endpoint).unwrap();
+    let requested_strata_address =
+        strata_address.map(|a| StrataAddress::from_str(&a).expect("bad strata address"));
+    let mut l1w =
+        SignetWallet::new(&seed, settings.network, settings.signet_backend.clone()).unwrap();
+    let l2w = StrataWallet::new(&seed, &settings.strata_endpoint).unwrap();
 
-    l1w.sync(&esplora).await.unwrap();
+    l1w.sync().await.unwrap();
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/deposit.rs:57]: Fix these `unwrap`s/`expect`s.

File: bin/strata-cli/src/cmd/deposit.rs [hunk 4]
```diff
@@ -65,11 +76,11 @@ pub async fn bridge_in(args: BridgeInArgs, seed: Seed, settings: Settings, esplo
...
         .clone()
-        .into_wallet_descriptor(l1w.secp_ctx(), NETWORK)
+        .into_wallet_descriptor(l1w.secp_ctx(), settings.network)
         .expect("valid descriptor");
 
     let mut temp_wallet = Wallet::create_single(desc.clone())
-        .network(NETWORK)
+        .network(settings.network)
         .create_wallet_no_persist()
         .expect("valid wallet");
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/deposit.rs:85]: Probably should fix these `expect`s.

File: bin/strata-cli/src/cmd/deposit.rs [hunk 5]
```diff
@@ -134,11 +141,17 @@ pub async fn bridge_in(args: BridgeInArgs, seed: Seed, settings: Settings, esplo
 
     let pb = ProgressBar::new_spinner().with_message("Broadcasting transaction");
     pb.enable_steady_tick(Duration::from_millis(100));
-    esplora.broadcast(&tx).await.expect("successful broadcast");
-    pb.finish_with_message(format!("Transaction {} broadcasted", tx.compute_txid()));
+    settings
+        .signet_backend
+        .broadcast_tx(&tx)
+        .await
+        .expect("successful broadcast");
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/deposit.rs:148]: Fix this `expect`.

File: bin/strata-cli/src/cmd/drain.rs [hunk 6]
```diff
@@ -63,22 +68,25 @@ pub async fn drain(
...
             .drain_to(address.script_pubkey())
-            .fee_rate(fr)
+            .fee_rate(fee_rate)
             .clone()
             .finish()
             .expect("valid transaction");
         l1w.sign(&mut psbt, Default::default()).unwrap();
         let tx = psbt.extract_tx().expect("fully signed tx");
-        esplora.broadcast(&tx).await.unwrap();
+        settings.signet_backend.broadcast_tx(&tx).await.unwrap();
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/drain.rs:84]: Fix these `unwrap`s/`expect`s.

File: bin/strata-cli/src/cmd/drain.rs [hunk 7]
```diff
@@ -46,15 +50,16 @@ pub async fn drain(
...
     });
     let strata_address =
         strata_address.map(|a| StrataAddress::from_str(&a).expect("valid Strata address"));
 
     if let Some(address) = signet_address {
-        let mut l1w = SignetWallet::new(&seed, NETWORK).unwrap();
-        l1w.sync(&esplora).await.unwrap();
+        let mut l1w =
+            SignetWallet::new(&seed, settings.network, settings.signet_backend.clone()).unwrap();
+        l1w.sync().await.unwrap();
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/drain.rs:62]: Fix these `unwrap`s.

File: bin/strata-cli/src/cmd/faucet.rs [hunk 8]
```diff
@@ -85,9 +92,12 @@ pub async fn faucet(args: FaucetArgs, seed: Seed, settings: Settings) {
         "✔ Solved challenge after {solution} attempts. Claiming now."
     ));
 
+    #[cfg(feature = "strata_faucet")]
     let url = match network_type {
         NetworkType::Signet => {
-            let mut l1w = SignetWallet::new(&seed, NETWORK).unwrap();
+            let mut l1w =
+                SignetWallet::new(&seed, settings.network, settings.signet_backend.clone())
+                    .unwrap();
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/faucet.rs:100]: Fix this `unwrap`.

File: bin/strata-cli/src/cmd/faucet.rs [hunk 9]
```diff
@@ -97,7 +107,7 @@ pub async fn faucet(args: FaucetArgs, seed: Seed, settings: Settings) {
                 Some(address) => {
                     let address = Address::from_str(&address).expect("bad address");
                     address
-                        .require_network(NETWORK)
+                        .require_network(settings.network)
                         .expect("wrong bitcoin network")
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/faucet.rs:111]: Fix this `expect`.

File: bin/strata-cli/src/cmd/faucet.rs [hunk 10]
```diff
@@ -126,11 +136,52 @@ pub async fn faucet(args: FaucetArgs, seed: Seed, settings: Settings) {
...
+            let _ = print_explorer_url(
+                &Txid::from_str(&body).expect("valid txid"),
+                &term,
+                &settings,
+            );
+        }
+        #[cfg(not(feature = "strata_faucet"))]
+        let _ = print_explorer_url(
+            &Txid::from_str(&body).expect("valid txid"),
+            &term,
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/faucet.rs:182]: Fix these `unwrap`s/`expect`s.

File: bin/strata-cli/src/cmd/receive.rs [hunk 11]
```diff
@@ -21,22 +20,24 @@ pub struct ReceiveArgs {
...
-            l1w.sync(&esplora).await.unwrap();
+            l1w.sync().await.unwrap();
             let _ = term.write_line("Wallet synced");
             let address_info = l1w.reveal_next_address(KeychainKind::External);
             l1w.persist().unwrap();
             address_info.address.to_string()
         }
         NetworkType::Strata => {
-            let l2w = StrataWallet::new(&seed, &settings.l2_http_endpoint).unwrap();
+            let l2w = StrataWallet::new(&seed, &settings.strata_endpoint).unwrap();
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/receive.rs:40]: Fix these `unwrap`s.

File: bin/strata-cli/src/cmd/recover.rs [hunk 12]
```diff
@@ -1,28 +1,31 @@
...
+}
 
-pub async fn refresh(seed: Seed, settings: Settings, esplora: EsploraClient) {
+pub async fn recover(args: RecoverArgs, seed: Seed, settings: Settings) {
     let term = Term::stdout();
-    let mut l1w = SignetWallet::new(&seed, NETWORK).unwrap();
-    l1w.sync(&esplora).await.unwrap();
+    let mut l1w =
+        SignetWallet::new(&seed, settings.network, settings.signet_backend.clone()).unwrap();
+    l1w.sync().await.unwrap();
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/recover.rs:28]: Fix this `unwrap`.

File: bin/strata-cli/src/cmd/recover.rs [hunk 13]
```diff
@@ -40,30 +43,36 @@ pub async fn refresh(seed: Seed, settings: Settings, esplora: EsploraClient) {
...
+        sync_wallet(&mut recovery_wallet, settings.signet_backend.clone())
+            .await
+            .expect("successful recovery wallet sync");
         let needs_recovery = recovery_wallet.balance().confirmed > Amount::ZERO;
 
         if !needs_recovery {
+            assert!(key.len() > 4);
+            let desc_height = u32::from_be_bytes(unsafe { *(key[..4].as_ptr() as *const [_; 4]) });
+            if desc_height + RECOVERY_DESC_CLEANUP_DELAY > current_height {
+                descriptor_file.remove(key).expect("removal should succeed");
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/recover.rs:71]: Fix these `expect`s.

File: bin/strata-cli/src/cmd/recover.rs [hunk 14]
```diff
@@ -92,9 +101,10 @@ pub async fn refresh(seed: Seed, settings: Settings, esplora: EsploraClient) {
...
 
         let tx = psbt.extract_tx().unwrap();
-        esplora
-            .broadcast(&tx)
+        settings
+            .signet_backend
+            .broadcast_tx(&tx)
             .await
-            .expect("successful tx broadcast");
+            .expect("broadcast to succeed")
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/recover.rs:108]: Fix this `expect`.

File: bin/strata-cli/src/cmd/reset.rs [hunk 15]
```diff
@@ -4,7 +4,8 @@ use dialoguer::Confirm;
 
 use crate::{seed::EncryptedSeedPersister, settings::Settings};
 
-/// Prints a BIP39 mnemonic encoding the internal wallet's seed bytes
+/// DANGER: resets the CLI completely, destroying all keys and databases.
+/// Keeps config.
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/reset.rs:8]: This warning should really be in the `--help` description shouldn't it?

File: bin/strata-cli/src/cmd/scan.rs [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+
+/// Performs a full scan of the signet wallet
+#[derive(FromArgs, PartialEq, Debug)]
+#[argh(subcommand, name = "scan")]
+pub struct ScanArgs {}
+
+pub async fn scan(_args: ScanArgs, seed: Seed, settings: Settings) {
+    let mut l1w =
+        SignetWallet::new(&seed, settings.network, settings.signet_backend.clone()).unwrap();
+    l1w.scan().await.unwrap();
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/scan.rs:13]: Fix these `unwrap`s.

File: bin/strata-cli/src/cmd/send.rs [hunk 17]
```diff
@@ -35,25 +34,29 @@ pub struct SendArgs {
...
-            let mut l1w = SignetWallet::new(&seed, NETWORK).expect("valid wallet");
-            l1w.sync(&esplora).await.unwrap();
-            let fee_rate = get_fee_rate(1, &esplora)
-                .await
-                .expect("valid response")
-                .expect("valid target");
+            let mut l1w =
+                SignetWallet::new(&seed, settings.network, settings.signet_backend.clone())
+                    .expect("valid wallet");
+            l1w.sync().await.unwrap();
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/send.rs:57]: Fix these `unwrap`s/`expect`s.

File: bin/strata-cli/src/cmd/send.rs [hunk 18]
```diff
@@ -64,28 +67,30 @@ pub async fn send(args: SendArgs, seed: Seed, settings: Settings, esplora: Esplo
...
-                .with_value(U256::from(args.amount * 10u64.pow(10)));
-            l2w.send_transaction(tx)
+                .with_value(U256::from(args.amount as u128 * SATS_TO_WEI));
+            let res = l2w
+                .send_transaction(tx)
                 .await
-                .expect("successful broadcast")
-                .tx_hash()
-                .0
+                .expect("successful broadcast");
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/send.rs:86]: Fix these `unwrap`s/`expect`s.

File: bin/strata-cli/src/cmd/withdraw.rs [hunk 19]
```diff
@@ -10,33 +10,34 @@ use console::Term;
...
-            .require_network(NETWORK)
+            .require_network(settings.network)
             .expect("correct network")
     });
 
-    let mut l1w = SignetWallet::new(&seed, NETWORK).unwrap();
-    let l2w = StrataWallet::new(&seed, &settings.l2_http_endpoint).unwrap();
+    let mut l1w =
+        SignetWallet::new(&seed, settings.network, settings.signet_backend.clone()).unwrap();
+    let l2w = StrataWallet::new(&seed, &settings.strata_endpoint).unwrap();
```
Comment (delbonis) [inline bin/strata-cli/src/cmd/withdraw.rs:40]: Fix these `unwrap`s/`expect`s.

File: bin/strata-cli/src/seed.rs [hunk 20]
```diff
@@ -27,44 +28,49 @@ impl BaseWallet {
...
+    fn gen<R: CryptoRng + RngCore>(rng: &mut R) -> Self {
+        let mut bytes = Zeroizing::new([0u8; SEED_LEN]);
+        rng.fill_bytes(bytes.as_mut());
+        Self(bytes)
     }
 
     pub fn print_mnemonic(&self, language: Language) {
         let term = Term::stdout();
-        let mnemonic = Mnemonic::from_entropy_in(language, &self.0).expect("valid entropy");
+        let mnemonic = Mnemonic::from_entropy_in(language, self.0.as_ref()).expect("valid entropy");
```
Comment (delbonis) [inline bin/strata-cli/src/seed.rs:42]: Fix this `expect`.

File: bin/strata-cli/src/seed.rs [hunk 21]
```diff
@@ -27,44 +28,49 @@ impl BaseWallet {
...
 
         let (salt_and_nonce, rest) = buf.split_at_mut(PW_SALT_LEN + AES_NONCE_LEN);
         let (seed, _) = rest.split_at_mut(SEED_LEN);
-        seed.copy_from_slice(&self.0);
+        seed.copy_from_slice(self.0.as_ref());
 
-        let mut cipher =
-            Aes256GcmSiv::new_from_slice(seed_encryption_key).expect("should be correct key size");
+        let mut cipher = Aes256GcmSiv::new_from_slice(seed_encryption_key.as_ref())
+            .expect("should be correct key size");
```
Comment (delbonis) [inline bin/strata-cli/src/seed.rs:73]: Fix this `expect`, if it could occur based on incorrect user input.

File: bin/strata-cli/src/seed.rs [hunk 22]
```diff
@@ -111,21 +117,27 @@ impl EncryptedSeed {
...
+            .seed_encryption_key(
+                &self.0[..PW_SALT_LEN].try_into().expect("cannot fail"),
+                HashVersion::V0,
+            )
             .map_err(OneOf::new)?;
 
-        let mut cipher =
-            Aes256GcmSiv::new_from_slice(seed_encryption_key).expect("should be correct key size");
+        let mut cipher = Aes256GcmSiv::new_from_slice(seed_encryption_key.as_ref())
+            .expect("should be correct key size");
```
Comment (delbonis) [inline bin/strata-cli/src/seed.rs:127]: Same as the other AES call.

File: bin/strata-cli/src/seed/password.rs [hunk 23]
```diff
@@ -1,45 +1,99 @@
...
-            None => {
-                let mut sek = [0u8; 32];
-                Argon2::default().hash_password_into(self.inner.as_bytes(), salt, &mut sek)?;
-                self.seed_encryption_key = Some(sek);
-                self.seed_encryption_key(salt)
+        version: HashVersion,
+    ) -> Result<Zeroizing<[u8; 32]>, argon2::Error> {
+        let mut sek = Zeroizing::new([0u8; 32]);
+        let (algo, ver, params) = version.params();
+        Argon2::new(algo, ver, params.expect("valid params")).hash_password_into(
```
Comment (delbonis) [inline bin/strata-cli/src/seed/password.rs:62]: Fix this `expect`.

File: bin/strata-cli/src/settings.rs [hunk 24]
```diff
@@ -3,77 +3,126 @@ use std::{
...
+                Arc::new(EsploraClient::new(&url).expect("valid esplora url"))
+            }
+            (None, Some(user), Some(pw), None, Some(url)) => Arc::new(Arc::new(
+                Client::new(&url, Auth::UserPass(user, pw)).expect("valid bitcoin core client"),
+            )),
+            (None, None, None, Some(cookie_file), Some(url)) => Arc::new(Arc::new(
+                Client::new(&url, Auth::CookieFile(cookie_file))
+                    .expect("valid bitcoin core client"),
+            )),
+            _ => panic!("invalid config for signet - configure for esplora or bitcoind"),
```
Comment (delbonis) [inline bin/strata-cli/src/settings.rs:101]: Fix these `expect`s and `panic!`.

File: bin/strata-cli/src/settings.rs [hunk 25]
```diff
@@ -3,77 +3,126 @@ use std::{
...
-                let mut buf = [0u8; 32];
-                decode(BRIDGE_MUSIG2_PUBKEY, &mut buf).expect("valid hex");
-                buf
+            faucet_endpoint: from_file.faucet_endpoint,
+            mempool_endpoint: from_file.mempool_endpoint,
+            bridge_musig2_pubkey: XOnlyPublicKey::from_slice(&match from_file.bridge_pubkey {
+                Some(key) => key.0,
+                None => {
+                    let mut buf = [0u8; 32];
+                    decode(BRIDGE_MUSIG2_PUBKEY, &mut buf).expect("valid hex");
```
Comment (delbonis) [inline bin/strata-cli/src/settings.rs:114]: Fix this `expect`.

File: bin/strata-cli/src/signet.rs [hunk 26]
```diff
@@ -90,131 +109,98 @@ impl SignetWallet {
                         .create_wallet(&mut Persister)
                         .expect("wallet creation to succeed")
```
Comment (delbonis) [inline bin/strata-cli/src/signet.rs:110]: Fix these `expect`s.

File: bin/strata-cli/src/signet.rs [hunk 27]
```diff
@@ -90,131 +109,98 @@ impl SignetWallet {
...
 
-impl DerefMut for SignetWallet {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        &mut self.0
-    }
+    apply_update_stream(wallet, rx).await;
+
+    handle
+        .await
+        .expect("thread to be fine")
```
Comment (delbonis) [inline bin/strata-cli/src/signet.rs:147]: Fix this `expect`.

File: bin/strata-cli/src/signet.rs [hunk 28]
```diff
@@ -90,131 +109,98 @@ impl SignetWallet {
...
-///
-/// Can only be set once - will return whether value was set.
-pub fn set_data_dir(data_dir: PathBuf) -> bool {
-    DATA_DIR.set(data_dir).is_ok()
-}
+    apply_update_stream(wallet, rx).await;
+
+    handle
+        .await
+        .expect("thread to be fine")
```
Comment (delbonis) [inline bin/strata-cli/src/signet.rs:167]: Fix this `expect`.

File: bin/strata-cli/src/signet.rs [hunk 29]
```diff
@@ -90,131 +109,98 @@ impl SignetWallet {
...
+            }
+            WalletUpdate::SpkScan(update) => {
+                wallet.apply_update(update).expect("update to connect")
+            }
+            WalletUpdate::NewBlock(ev) => {
+                let height = ev.block_height();
+                let connected_to = ev.connected_to();
+                wallet
+                    .apply_block_connected_to(&ev.block, height, connected_to)
+                    .expect("block to be added")
```
Comment (delbonis) [inline bin/strata-cli/src/signet.rs:187]: Fix these `expect`s.

File: bin/strata-cli/src/signet/backend.rs [hunk 30]
```diff
@@ -0,0 +1,328 @@
...
+            "Applied {} unconfirmed transactions in {:?}",
+            txs_len, elapsed
+        ));
+        Ok(())
+    })
+    .await
+}
+
+#[derive(Debug)]
+pub struct InvalidFee;
```
Comment (delbonis) [inline bin/strata-cli/src/signet/backend.rs:328]: Should split error types out into their own module, ideally.

File: bin/strata-cli/src/signet/persist.rs [hunk 31]
```diff
@@ -0,0 +1,63 @@
...
+/// Must be called before accessing [`Persister`].
+///
+/// Can only be set once - will return whether value was set.
+pub fn set_data_dir(data_dir: PathBuf) -> bool {
+    DATA_DIR.set(data_dir).is_ok()
+}
+
+thread_local! {
+    static DB: Rc<RefCell<Connection>> = RefCell::new(Connection::open(SignetWallet::db_path("default", DATA_DIR.get().expect("data dir to be set"))).unwrap()).into();
+}
```
Comment (delbonis) [inline bin/strata-cli/src/signet/persist.rs:32]: This data should be passed around as struct fields along with the data associated with it, not in globals/tls.

**Review by delbonis — COMMENTED** [review]
Panics should only be used when they'd be triggered as of an actual bug in our code, an assumption being violated.  *Not* something that can be caused by some normal user error.  See [ticket](https://alpenlabs.atlassian.net/browse/STR-546).

**Comment by storopoli** [comment]
@delbonis can you check #437? that was the one we merged.

**Pull request closed without merging**